File: JS/src/blockViewController.js
_import(
  "NSMutableDictionary,NSMutableArray,UIView,UILabel,UIImage,UIColor,TTView,ViewController,UITableViewCell,UITableView,NSIndexPath,UIFont,UIScreen,UIImageView,TaoBaoHome"
);

defineClass(
  "BlockViewController:UITableViewController",
  {
    dealloc: function () {
      Utils.log("BlockViewController->已释放");
    },
    viewDidLoad: function () {
      Super().viewDidLoad();
      self.addSomeTestView();
    },
    refresh: function () {
      // Super().testSuper();
      // self.testSuper();
      self.cleanSubviews();
      self.addSomeTestView();
    },
    cleanSubviews: function () {
      self
        .view()
        .subviews()
        .forEach((subview) => {
          if (subview != "UITableView") {
            subview.removeFromSuperview();
          } else {
            self.tableView().reloadData();
          }
        });
    },
    sendMessage_: function (msg) {
      Utils.log_info(msg);
    },
    sendMessageVC_: function (vc) {
      vc.view().setBackgroundColor_(UIColor.blackColor());
    },
    addSomeTestView: function () {
      self.setTitle_("BlockViewController");
      self.view().setBackgroundColor_(UIColor.whiteColor());
      let screenWidth = UIScreen.mainScreen().bounds().size.width;
      let screenHeight = UIScreen.mainScreen().bounds().size.height;

      let logo = UIImageView.new();
      logo.setImage_(UIImage.imageNamed_("applelogo"));
      logo.setFrame_(new TTReact(50, 50, 100, 100));
      logo.setCenter_(new TTPoint(screenWidth / 2, screenHeight - 250));
      let title = UILabel.new();
      title.setText_("Apple");
      title.setFont_(UIFont.fontWithName_size_("GillSans-UltraBold", 25));
      title.setTextAlignment_(1);
      title.setFrame_(new TTReact(50, 150, 150, 100));
      title.setCenter_(new TTPoint(screenWidth / 2, screenHeight - 150));
      self.view().addSubview_(logo);
      self.view().addSubview_(title);
    },
    btnAction_: function (index) {
      switch (index) {
        case 0:
          {
            self.noReturnParamsVoid_(
              block("void,void", function () {
                Utils.log_info(
                  "--------JS传入OC方法,接受到回调--------- 无参数,无返回值"
                );
              })
            );
          }
          break;
        case 1:
          {
            self.noReturnParamsStringInt_(
              block("void,id,int", function (arg1, arg2) {
                var dic = JSON.parse(arg1);
                Utils.log_info(
                  "--------JS传入OC方法,接受到回调--------- 有参数,无返回值  " +
                    dic.name +
                    "-" +
                    arg2
                );
              })
            );
          }
          break;
        case 2:
          {
            Utils.log_info("runBlock, js调用");
            self.returnParamsId_(
              block("id,id", function (arg) {
                var dic = JSON.parse(
                  '{"id":1,"name":"jb51","email":"admin@jb51.net","interest":["wordpress","php"]}'
                );
                Utils.log_info(
                  "--------JS传入OC方法,接受到回调--------- 有参数,有返回值:string  " +
                    arg
                );
                arg.view().setBackgroundColor_(UIColor.blackColor());
                return dic;
              })
            );
          }
          break;
        case 3:
          {
            Utils.log_info("runBlock, js调用");
            self.runBlock();
          }
          break;

        default:
          {
            //方法签名第一位 是返回值,如果返回为void可以不填,但是要以","分割
            self.testMoreParams_(
              block("id, id, id, int, bool, float, id", function (
                arg1,
                arg2,
                arg3,
                arg4,
                arg5,
                arg6
              ) {
                Utils.log_info(
                  "--------JS传入OC方法,接受到回调---------" +
                    arg1 +
                    "\n" +
                    arg2 +
                    "\n" +
                    arg3 +
                    "\n" +
                    arg4 +
                    "\n" +
                    arg5 +
                    "\n" +
                    arg6
                );
              })
            );
            // self.OCcallBlock_(
            //   block(",id", function (arg1) {
            //     Utils.log_info("js与js block回调" + arg1);
            //   })
            // );
          }
          break;
      }
    },
    OCNoReturnParams_: function (callback) {
      Utils.log_info("OC block 传入JS, js调用");
      if (callback) {
        callback("OCNoReturnParams: arg: 10");
      }
    },
  },
  {}
);

File: TTDFKitExample/TTDFKitExample/实际场景/BlockViewController.m
//
//  BlockViewController.m
//  Example
//
//  Created by tianyubing on 2019/9/7.
//  Copyright © 2019 TianyuBing. All rights reserved.
//

#import "BlockViewController.h"
@interface TTPlaygroundModel : NSObject
@property(nonatomic,strong)NSString *name;
@end

@implementation TTPlaygroundModel



@end


#define data @[@"1.JS替换:noReturnParamsVoid, 无返回/无参 ",\
                @"2.JS替换:noReturnParamsStringInt, 无返回/Str,Int",\
                @"3.JS替换:returnParamsId, 有返回/有参",\
                @"4.JS调用OC:OCNoReturnParams, 有返回/有参",\
                @"5.JS替换:testMoreParams, 多参数&多类型"]

@interface BlockViewController ()

@end

@implementation BlockViewController



- (void)viewDidLoad {
    [super viewDidLoad];
 
    // Do any additional setup after loading the view.
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section{
    UILabel *msg = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width, 100)];
    msg.text = @"    参数携带block参数请参考例:1~4.\n    常规方法替换请参照[testMoreParams]";
    msg.numberOfLines = 0;
    msg.backgroundColor = [UIColor systemGreenColor];
    msg.textColor = [UIColor whiteColor];
    return msg;
}

- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section{
    return 100;;
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return [data count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    UITableViewCell * cell = [[UITableViewCell alloc] initWithStyle:(UITableViewCellStyleDefault) reuseIdentifier:@"cell"];
    cell.textLabel.text = data[indexPath.row];
    return cell;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    [self btnAction:(int)indexPath.row];
}

#pragma mark - 演示代码

- (void)noReturnParamsVoid:(void(^)(void))call{
    if (call) {
        call();
    }
    [self sendMessage:@"1,2,3"];
    [self sendMessageVC:self];
}

- (void)noReturnParamsStringInt:(void(^)(NSString * str,int inta))call{
    if (call) {
        call(@"{\"id\":1,\"name\":\"Tencent\",\"email\":\"admin@Tencent.com\",\"interest\":[\"Tencent\",\"Tencent\"]}",999);
    }
}

- (void)returnParamsId:(NSDictionary *(^)(UIViewController *str))call{
    if (call) {
        NSDictionary *  res = call(self);
        NSLog(@"有参block----%@",res);
    }
}



- (void)testMoreParams:(void(^)(NSString *arg0,NSString *arg1,int arg2,bool arg3, float arg4, NSNumber* arg5))call{
    if (call) {
        call(@"arg0",@"arg1 ",24,NO,1.99,@(58));
    }else{
        NSLog(@"--------Call 方法未实现---------");
    }
}


- (void)OCcallBlock:(void(^)(NSString *str))block{
    if (block) {
        block(@"第二个block");
    }
}

- (void)OCNoReturnParams:(void(^)(NSString *str))block{
    
}

- (void)runBlock{
    id cbInt = ^void(NSString *arg) {
        NSLog(@"js方法回调----------%@", arg);
    };
    
    [self OCNoReturnParams:cbInt]; //it's OK
}


- (void)configViewSize:(CGSize )size{
    
}

- (void)configView:(UIView * )view{
    
}


- (void)callBlock:(void(^)(NSString *str))block
           block2:(void(^)(NSString *str))block2
{
    
}

- (void)btnAction:(int)index{
    
}

- (void)testCall3:(NSString *(^)(void))call{
    if (call) {
        NSLog(@"有参block----%@",call());
    }
}


//- (void)invocateTest{
//    NSLog(@"开始调用");
//    [self testCallVV:^{
//        NSLog(@"接受回调");
//    }];
//
//    [self testCallIDID:^TTPlaygroundModel *(NSString *str) {
//        NSLog(@"接受回调 -- %@",str);
//        TTPlaygroundModel *model = [TTPlaygroundModel new];
//        model.name = @"TTDFKit";
//        return model;
//    }];
//}

@end
